# Exercices XP : Cahier d'exercices guidé

Ce cahier suit le **contenu de la plateforme**. Certaines cellules sont **pré-remplies** et doivent être **uniquement exécutées**. Les cellules marquées **To-Do** nécessitent votre action. Lorsqu'une réponse écrite est requise, le **To-Do** apparaît dans une cellule Markdown. Lorsqu'un code est requis, le **To-Do** apparaît dans une cellule de code sous forme de commentaires.

Les **points d'apprentissage** apparaissent lorsqu'un concept est important pour l'intuition ou pour vous aider à apprendre de nouvelles astuces/concepts liés à d'autres sujets de l'IA.

In [ ]:
pip install tensorflow

## Objectifs

**Ce que vous allez apprendre**  
- Les fondamentaux du deep learning et des réseaux de neurones.
- Comment construire et entraîner des réseaux de neurones simples avec TensorFlow ou Keras.
- Les concepts de propagation avant (forward) et arrière (backward).
- Comment visualiser et interpréter les prédictions d'un modèle.

**Ce que vous allez créer**  
- Un système de décision simple basé sur un perceptron.
- Un réseau de neurones pour classer les chiffres manuscrits du jeu de données MNIST.
- Un calcul de propagation avant pour prédire le prix des maisons.
- Une implémentation Python de la propagation avant et arrière.
- Des visualisations des prédictions faites par un réseau de neurones sur le jeu de données MNIST.

## 🌟 Exercice 1 : Petit Quiz

- **Quelle est la différence clé entre l'apprentissage automatique (machine learning) traditionnel et le deep learning ?**
L'apprentissage automatique traditionnel nécessite souvent une ingénierie manuelle des caractéristiques (feature engineering) où des experts identifient les attributs pertinents, tandis que le deep learning apprend automatiquement des caractéristiques hiérarchiques à partir de données brutes à l'aide de réseaux de neurones.
- **Comment les réseaux de neurones artificiels imitent-ils le cerveau humain ?**
Ils imitent le cerveau grâce à une structure de nœuds interconnectés (neurones) qui traitent l'information par couches, transmettant des signaux via des poids et des fonctions d'activation similaires aux synapses biologiques.
- **Pourquoi le deep learning est-il plus performant sur de grands ensembles de données que l'apprentissage automatique traditionnel ?**
Les modèles de deep learning ont une capacité et une complexité plus élevées, ce qui leur permet de capturer des modèles complexes dans de vastes quantités de données sans voir leur performance 'stagner', contrairement aux modèles traditionnels.
- **Quels sont les défis du deep learning et comment peuvent-ils être relevés ?**
Les défis incluent un coût de calcul élevé (résolu par les GPU/TPU) et le besoin de grands ensembles de données étiquetées (résolu par l'augmentation de données ou le transfert d'apprentissage).
- **Quel rôle jouent les couches cachées dans un modèle de deep learning ?**
Les couches cachées agissent comme des extracteurs de caractéristiques qui transforment les données d'entrée en représentations de plus en plus abstraites, permettant au modèle d'apprendre des relations non linéaires.
- **Dans un réseau de neurones artificiels, quelle est la fonction d'une fonction d'activation ?**
Une fonction d'activation introduit de la non-linéarité dans le réseau, lui permettant d'apprendre et de représenter des modèles complexes qu'un modèle purement linéaire ne pourrait pas saisir.

## 🌟 Exercice 2 : Création d'un système de décision par perceptron simple

Vous allez créer un système de décision basé sur un perceptron pour décider si vous devez sortir en fonction de deux entrées :
- Température en °F
- Pluie (Oui = 1, Non = 0)

Paramètres donnés : Poids température 0.6, Poids pluie 0.4, Biais 2. Calculez la somme pondérée :
\[ \text{Somme pondérée} = 0.6 \cdot \text{Température} + 0.4 \cdot \text{Pluie} + \text{Biais} \]
Appliquez une activation par palier (step activation) :
- Si Somme pondérée > 20, sortie = 1
- Si Somme pondérée ≤ 20, sortie = 0

Évaluez :
Cas 1 : Température 70°F, Pluie 0
Cas 2 : Température 50°F, Pluie 1

In [ ]:
# PRÉ-REMPLI : exécutez simplement
from typing import Tuple

W_TEMP = 0.6
W_RAIN = 0.4
BIAS   = 2.0

def step_activation(s: float, threshold: float = 20.0) -> int:
    """Retourne 1 si s > threshold sinon 0."""
    return 1 if s > threshold else 0

In [ ]:
# Définition des constantes
W_TEMP = 0.6
W_RAIN = 0.4
BIAS = 2.0

def weighted_sum(temperature_f: float, rain01: int) -> float:
    """Calcule s = 0.6*Température + 0.4*Pluie + Biais."""
    return (W_TEMP * temperature_f) + (W_RAIN * rain01) + BIAS

def step_activation(s: float, threshold: float = 20.0) -> int:
    """Retourne 1 si s > seuil sinon 0."""
    return 1 if s > threshold else 0

# Calcul des décisions pour les deux cas
case1 = (70, 0)
case2 = (50, 1)

s1 = weighted_sum(*case1)
s2 = weighted_sum(*case2)
y1 = step_activation(s1)
y2 = step_activation(s2)

print({"cas 1": {"somme": s1, "décision": y1}, "cas 2": {"somme": s2, "décision": y2}})

{'cas 1': {'somme': 44.0, 'décision': 1}, 'cas 2': {'somme': 32.4, 'décision': 1}}


**Interprétation :**
- **Cas 1 (70°F, Pas de pluie) :** La somme pondérée est 44.0 (0.6*70 + 0 + 2). Comme 44.0 > 20, la décision est 1 (Sortir).
- **Cas 2 (50°F, Pluie) :** La somme pondérée est 32.4 (0.6*50 + 0.4*1 + 2). Comme 32.4 > 20, la décision est 1 (Sortir).
Dans les deux cas, la température élevée ou le biais positif ont maintenu la somme au-dessus du seuil de 20, suggérant de sortir même lorsqu'il pleuvait dans le Cas 2.

## 🌟 Exercice 3 : Construire un réseau de neurones simple avec TensorFlow ou Keras

Créez un classificateur pour MNIST avec :
- Une couche d'entrée
- Une couche cachée avec 128 neurones et une activation ReLU
- Une couche de sortie avec 10 neurones et une activation softmax

Étapes :
1. Importer TensorFlow et les modules nécessaires
2. Charger le jeu de données MNIST
3. Normaliser les données (diviser par 255.0)
4. Encoder les étiquettes en 'one-hot'
5. Construire un modèle séquentiel (Flatten, Dense 128 ReLU, Dense 10 Softmax)
6. Compiler avec l'optimiseur `adam`, la perte `categorical_crossentropy` et la métrique `accuracy`
7. Entraîner et évaluer sur l'ensemble de test

In [ ]:
# PREFILLED: just execute
import numpy as np
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.datasets import mnist

print("TensorFlow:", tf.__version__)

# Load data
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Normalize to [0,1]
x_train = x_train.astype("float32") / 255.0
x_test  = x_test.astype("float32") / 255.0

# Flatten will be added in the model; keep images as 28x28 for now
print("Train shape:", x_train.shape, "Test shape:", x_test.shape)


TensorFlow: 2.20.0
Train shape: (60000, 28, 28) Test shape: (10000, 28, 28)


In [ ]:
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.datasets import mnist

# Chargement et normalisation des données localement pour garantir l'existence des variables
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.astype("float32") / 255.0
x_test  = x_test.astype("float32") / 255.0

# Encodage One-hot des étiquettes
y_train_oh = to_categorical(y_train, 10)
y_test_oh  = to_categorical(y_test, 10)

print("Forme des étiquettes one-hot :", y_train_oh.shape, y_test_oh.shape)

Forme des étiquettes one-hot : (60000, 10) (10000, 10)


In [ ]:
from tensorflow.keras import layers, models

# Construction du modèle
model = models.Sequential([
    layers.Flatten(input_shape=(28, 28)),
    layers.Dense(128, activation='relu'),
    layers.Dense(10, activation='softmax'),
])

model.summary()

/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_2 (Flatten)             │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 101,770 (397.54 KB)

 Trainable params: 101,770 (397.54 KB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [ ]:
# Compilation, Entraînement et évaluation
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
history = model.fit(x_train, y_train_oh, epochs=5, batch_size=32, validation_split=0.1)
test_loss, test_acc = model.evaluate(x_test, y_test_oh)
print({"perte_test": float(test_loss), "precision_test": float(test_acc)})

Epoch 1/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.9203 - loss: 0.2763 - val_accuracy: 0.9625 - val_loss: 0.1317
Epoch 2/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9639 - loss: 0.1216 - val_accuracy: 0.9743 - val_loss: 0.0948
Epoch 3/5
1675/1688 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9748 - loss: 0.0853

## 🌟 Exercice 4 : Calcul de la propagation avant

Prédisez le prix d'une maison avec :
- Entrées : x₁ = 2000 pieds carrés, x₂ = 3 chambres
- Poids : w₁ = 0.5, w₂ = 0.7
- Biais : b = 50 000
- Activation : ReLU

Calculez z avant activation, appliquez ReLU et interprétez la prédiction.

In [ ]:
x1, x2 = 2000.0, 3.0
w1, w2 = 0.5, 0.7
b = 50000.0

z = x1*w1 + x2*w2 + b

# Application de ReLU
y = max(0, z)

print({"z": z, "prédiction": y})

**Interprétation :** Le prix prédit de la maison est de 51 002,10 $. Le biais (50 000) représente le prix de base d'une maison quels que soient sa taille ou ses chambres. La mise à l'échelle des entrées est cruciale car la surface habitable (2000) a un impact beaucoup plus important sur la somme finale que le nombre de chambres (3) compte tenu de leurs poids respectifs.

## 🌟 Exercice 5 (optionnel) : Implémentation de la propagation avant et arrière en Python

Vous allez coder un réseau de neurones simple qui effectue une propagation avant et une rétropropagation pour un problème de régression prédisant les scores d'examen à partir des heures d'étude.

In [ ]:
# PRÉ-REMPLI : exécutez simplement
import numpy as np

# Initialisation des données d'entrée
x = np.array([4, 80])  # 4 heures d'étude, score précédent : 80

# Initialisation des poids et du biais
w = np.array([0.6, 0.3])
b = 10

# Propagation avant
def forward_propagation(x, w, b):
    z = np.dot(x, w) + b
    return z

# Calcul de la prédiction
y_pred = forward_propagation(x, w, b)
y_true = 85  # Score réel

# Calcul de la perte (Erreur Quadratique Moyenne)
loss = 0.5 * (y_true - y_pred) ** 2

# Calcul des gradients
grad_w = -(y_true - y_pred) * x
grad_b = -(y_true - y_pred)

# Mise à jour des poids et du biais
learning_rate = 0.01
w_new = w - learning_rate * grad_w
b_new = b - learning_rate * grad_b

print("Prédiction initiale :", y_pred)
print("Perte :", loss)
print("Nouveaux poids :", w_new)
print("Nouveau biais :", b_new)

**Explication de la rétropropagation :** La descente de gradient réduit l'erreur en calculant la pente (gradient) de la fonction de perte par rapport à chaque poids. En déplaçant les poids dans la direction opposée au gradient, nous 'descendons' la courbe de perte vers l'erreur minimale.

**Expérience :** Si nous augmentons le taux d'apprentissage à 0.1, les mises à jour deviennent beaucoup plus importantes, ce qui peut aider à atteindre la solution plus rapidement mais risque de dépasser l'optimum.

## 🌟 Exercice 6 (optionnel) : Visualisation des prédictions sur MNIST

Entraînez un modèle simple sur MNIST, puis visualisez les prédictions pour certaines images de test avec les étiquettes prédites.

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

# Réutilisation des données et du modèle
idx = np.random.choice(len(x_test), 12)
x_vis = x_test[idx]
y_true_labels = y_test[idx]

y_prob = model.predict(x_vis)
y_pred = np.argmax(y_prob, axis=1)

plt.figure(figsize=(12, 5))
for i, k in enumerate(idx):
    plt.subplot(2, 6, i+1)
    plt.imshow(x_test[k], cmap='gray')
    plt.title(f'P:{y_pred[i]} T:{y_true_labels[i]}')
    plt.axis('off')
plt.tight_layout()
plt.show()